x-pack/plugins/alerting/server/types.ts

```diff
@@ -74,13 +74,14 @@ export interface RuleExecutorServices<
   Context extends AlertInstanceContext = AlertInstanceContext,
   ActionGroupIds extends string = never
 > {
-  searchSourceClient: ISearchStartSearchSource;
+  alertFactory: PublicAlertFactory<State, Context, ActionGroupIds>;
   savedObjectsClient: SavedObjectsClientContract;
-  uiSettingsClient: IUiSettingsClient;
   scopedClusterClient: IScopedClusterClient;
-  alertFactory: PublicAlertFactory<State, Context, ActionGroupIds>;
-  shouldWriteAlerts: () => boolean;
+  searchSourceClient: ISearchStartSearchSource;
+  uiSettingsClient: IUiSettingsClient;
+  getAlertUuid?: (alertId: string) => string | null;
```

Comment:
I think this is the only change.  But it's not appropriate, since the alerting framework doesn't know anything about alert UUIDs yet - though it will soon - https://github.com/elastic/kibana/pull/143489.

So, it really shouldn't be added here.  And I don't think this would work anyway, since there's nothing hooking this method up to the lifecycle executor version, so I believe the method will always be `undefined`.

If you can wait for the PR I mentioned to be merged (should be by 8.6 FF), you could wait till we add it.  Otherwise, we'll have to find some other way for you to access it.  I'd guess security is doing it somewhere so we can probably find a nice way to do it.

Reply:
@pmuellr that's surprising as I found it working (see the second screenshot in the PR description). 

I worked with @simianhacker to add this commit: https://github.com/elastic/kibana/pull/142854/commits/30367345952d9a7602a94faea7bdbc8ed63b4558#diff-cb5b88a0f28b67a629a54e0101adf980bd20b90812e042f65e59082316cc1570 

That added the alert UUID to the executor, even the first time the executor executed. This approach worked for us to gather the alert UUID for different rule types (Infra, Uptime, Logs).

At the same time, I am a newcomer in the codebase so if APM rules work different and you have something better in the works then I am totally willing to reverse this and use a solution that you can provide. I do however want to be certain this change reaches 8.6.0. Think that's possible?

Reply:
I think the `services` within `x-pack/plugins/apm/server/routes/alerts/rule_types/anomaly/register_anomaly_rule_type.ts` need to be typed to [`LifecycleAlertServices`](https://github.com/elastic/kibana/blob/main/x-pack/plugins/rule_registry/server/utils/create_lifecycle_executor.ts#L66) as `getAlertUuid` is defined within the rule registry (middle layer between O11y rule types and the framework).

Reply:
Ah, ya I think Mike hit on it ^^^.   

Thinking the reason the test works is because it's actually executing the rule registry lifecycle version, not the alerting framework version.

Reply:
Hey Mike and Patrick, thanks for your comments.

I've tried typing the services object in https://github.com/elastic/kibana/blob/main/x-pack/plugins/alerting/server/types.ts#L98 to `LifecycleAlertServices<InstanceState, InstanceContext, ActionGroupIds>` using the import `import type { LifecycleAlertServices } from '@kbn/rule-registry-plugin/server/utils/create_lifecycle_executor';`, but then `savedObjectsClient` and `scopedClusterClient` can't be found on the `services` object: 
<img width="586" alt="Screenshot 2022-11-09 at 09 58 29" src="https://user-images.githubusercontent.com/535564/200786516-29876b1e-782f-4078-ae4d-efccbd8cb131.png">

Going to try a few things, might reach out if I'm not successful. Thanks again.

Reply:
Looks like what gets sent is a combination of the base alerting services + the lifecycle services.  Not sure what _that_ type is though.  https://github.com/elastic/kibana/blob/7ac391c8c48d6f4bd9c9f1aed2b02471d74a3ba2/x-pack/plugins/rule_registry/server/utils/create_lifecycle_executor.ts#L187-L194

Reply:
The typing of the APM executor functions is different from the others, but I haven't figured out where this is set. 

Take for instance the executor function for the Infra > Inventory Metric Threshold rule type: https://github.com/elastic/kibana/blob/main/x-pack/plugins/infra/server/lib/alerting/inventory_metric_threshold/inventory_metric_threshold_executor.ts#L69

The typing of `services` there goes to 
```
export type AlertExecutorOptionsWithExtraServices<
  Params extends RuleTypeParams = never,
  State extends RuleTypeState = never,
  InstanceState extends AlertInstanceState = never,
  InstanceContext extends AlertInstanceContext = never,
  ActionGroupIds extends string = never,
  TExtraServices extends {} = never
> = Omit<
  RuleExecutorOptions<Params, State, InstanceState, InstanceContext, ActionGroupIds>,
  'services'
> & {
  services: RuleExecutorServices<InstanceState, InstanceContext, ActionGroupIds> & TExtraServices;
};
```

Compared to the APM transaction error threshold rule: https://github.com/elastic/kibana/blob/main/x-pack/plugins/apm/server/routes/alerts/rule_types/error_count/register_error_count_rule_type.ts#L98.

There services points to https://github.com/elastic/kibana/blob/main/x-pack/plugins/alerting/server/types.ts:98

This seems an error with typing, as the contents of both `services` objects during runtime seems the same.

Reply:
Did some further investigating.

All rule types of APM and Uptime are registered by using the `createLifecycleRuleType` function: 

```
  const createLifecycleRuleType = createLifecycleRuleTypeFactory({
    ruleDataClient,
    logger,
  });


registerType(createLifecycleRuleType(rule));

```

For example: https://github.com/elastic/kibana/blob/main/x-pack/plugins/apm/server/routes/alerts/rule_types/anomaly/register_anomaly_rule_type.ts#L84


The rule types for Infra and Logs are registered by using:

```
registerType(rule)

```
For example: https://github.com/elastic/kibana/blob/main/x-pack/plugins/infra/server/lib/alerting/log_threshold/register_log_threshold_rule_type.ts#L103, https://github.com/elastic/kibana/blob/main/x-pack/plugins/infra/server/lib/alerting/inventory_metric_threshold/register_inventory_metric_threshold_rule_type.ts#L82

Can anybody provide context as to why these different approaches exist?